std-bits/table/src/main/java/org/enso/table/data/column/storage/BoolStorage.java

```diff
@@ -402,30 +428,42 @@ public BoolStorage runZip(
                   BoolStorage storage,
                   Storage<?> arg,
                   MapOperationProblemAggregator problemAggregator) {
-                if (arg instanceof BoolStorage v) {
-                  BitSet missing = v.isMissing.get(0, storage.size);
-                  missing.or(storage.isMissing);
-                  BitSet out = v.values.get(0, storage.size);
-                  boolean negated;
-                  if (storage.negated && v.negated) {
-                    out.and(storage.values);
-                    negated = true;
-                  } else if (storage.negated) {
-                    out.flip(0, storage.size);
-                    out.and(storage.values);
-                    negated = true;
-                  } else if (v.negated) {
-                    out.flip(0, storage.size);
-                    out.or(storage.values);
-                    negated = false;
-                  } else {
-                    out.or(storage.values);
-                    negated = false;
-                  }
-                  return new BoolStorage(out, missing, storage.size, negated);
-                } else {
+                if (!(arg instanceof BoolStorage v)) {
                   throw new UnexpectedColumnTypeException("Boolean");
                 }
+
+                BitSet out = v.values.get(0, storage.size);
+                boolean negated;
+                if (storage.negated && v.negated) {
+                  out.and(storage.values);
+                  negated = true;
+                } else if (storage.negated) {
+                  out.flip(0, storage.size);
+                  out.and(storage.values);
+                  negated = true;
+                } else if (v.negated) {
+                  out.flip(0, storage.size);
+                  out.or(storage.values);
+                  negated = false;
+                } else {
+                  out.or(storage.values);
+                  negated = false;
+                }
+
+                BitSet missing = v.isMissing.get(0, storage.size);
+                missing.or(storage.isMissing);
+                int current = missing.nextSetBit(0);
+                while (current != -1) {
+                  var value = negated != out.get(current);
+                  if (value
+                      && (storage.getItemBoxed(current) == Boolean.TRUE
+                          || v.getItemBoxed(current) == Boolean.TRUE)) {
+                    missing.clear(current);
+                  }
+                  current = missing.nextSetBit(current + 1);
+                }
+
+                return new BoolStorage(out, missing, storage.size, negated);
```

Comment:
This is getting quite complicated (I admit it was complicated before too, but when it was my own code, I didn't feel that as much since I remembered the context better - now coming to it after a while it makes it a bit harder).

I suggest maybe we should split the And and Or operations into separate files? As the `BoolStorage.java` is quite large already.

Then, I'm wondering if it may be good to add some explanations in the comments what each piece is doing, e.g. I assume the `nextSetBit` loop is used to un-mark the part that is no longer undefined, because `true || null` made it into a proper value.

Reply:
I agree but would rather leave for a wider refactor of how we do Storage and Operations.